Prometheus: Streaming JSON parser performance improvements

toddtreece opened the pull request:
This adds the `prometheusWideSeries` feature flag, which enables matrix and vector wide series responses.

### `main` with `prometheusStreamingJSONParser` disabled (old) vs this branch (new):
```
name     old time/op    new time/op    delta
Json-32    78.0ms ±11%    32.2ms ± 9%  -58.71%  (p=0.000 n=10+10)

name     old alloc/op   new alloc/op   delta
Json-32    41.3MB ± 0%     7.1MB ± 0%  -82.72%  (p=0.000 n=10+10)

name     old allocs/op  new allocs/op  delta
Json-32      754k ± 0%      371k ± 0%  -50.73%  (p=0.000 n=10+9)
```

### `main` with `prometheusStreamingJSONParser` enabled (old) vs this branch (new):
```
name     old time/op    new time/op    delta
Json-32    36.2ms ± 6%    32.2ms ± 9%  -10.95%  (p=0.000 n=10+10)

name     old alloc/op   new alloc/op   delta
Json-32    19.7MB ± 0%     7.1MB ± 0%  -63.78%  (p=0.000 n=10+10)

name     old allocs/op  new allocs/op  delta
Json-32      389k ± 0%      371k ± 0%   -4.51%  (p=0.000 n=7+9)
```

toddtreece committed:
Prometheus: Move existing query logic to new package

toddtreece committed:
fix tests

toddtreece committed:
fix lint error

toddtreece committed:
delete unused types file

toddtreece committed:
move testdata back to original location

toddtreece committed:
Prometheus: Add query package

toddtreece committed:
Prometheus: Add streaming package

toddtreece committed:
add streaming bench test

toddtreece committed:
fix lint errors

toddtreece committed:
fix multi-query requests

toddtreece committed:
rename query -> models

toddtreece committed:
rename streaming -> querydata

toddtreece committed:
update tests

toddtreece committed:
fix path to client

toddtreece committed:
add missed test files

toddtreece committed:
Prometheus: Streaming JSON performance improvements

ryantxu commented:
If I understand the change, we are making the "matrix" call return "timeseries-wide" rather than many.  If that is accurate, we need to also update the return type.

perf improvements look good 🎉 also means the frontend does not need to join!

toddtreece committed:
switch to time series wide type for matrix & vector responses

toddtreece commented:
@ryantxu yep, for matrix and vector responses. good catch on the type. i updated that in [87bab64](https://github.com/grafana/grafana/pull/48792/commits/87bab64742284bdb01081109bd6b0adb7eb7c3d1)

i also switched to `[]*float64` for the value fields since wide frames will require `nil` to fill gaps

ryantxu commented:
> since wide frames will require nil to fill gaps

Do you know if the "gaps" are meaningful or not?

In the frontend we have an odd distinction between null and undefined -- we stick `undefined` values in places where the value is missing only as a byproduct of a join function.

I think we prototyped encodeing this as another variation of NaN before....

toddtreece commented:
@leeoniya would need to confirm, but i think the frontend uses `null` when joining now

ryantxu commented:
* null = missing, and will show a gap that will respect the interval (if set)
* undefined = value that is only missing because we were forced to join.

ryantxu commented:
This is a behavior change (i think good!) but we may want to wait on it until after switching from `buffered` to the new parser.

I think we *will* want this to be a changelog item, but only after we are using the new parser.

toddtreece commented:
@ryantxu we could make this change backwards compatible by setting a value only on new queries that indicate that they should return wide frames. existing and edited queries wouldn't have this property, so they would return multi/many. thoughts?

toddtreece commented:
> null = missing, and will show a gap that will respect the interval (if set)
> undefined = value that is only missing because we were forced to join.

@ryantxu i think that's true for some data sources, but multiple matrix series returned by prometheus share the same interval, so i believe any gaps caused by missing points will be null filled. this used to happen in the prometheus data source frontend code, and then @leeoniya added this https://github.com/grafana/grafana/pull/44622 when interval was added to the data frame

toddtreece commented:
@gabor is this change going to be an issue for loki, or does it behave the same as prometheus?

leeoniya commented:
> so i believe any gaps caused by missing points will be null filled

the renderer carries forward `null` values as gaps, so the null insertion is only done at the beginning of any run of adjacent nulls.

https://github.com/grafana/grafana/blob/454e8046573752d67371ca7493c211f762133de3/packages/grafana-ui/src/components/GraphNG/nullInsertThreshold.test.ts#L84-L98

gabor commented:
@toddtreece in general i do like the change, i think it's the way forward 👍

toddtreece commented:
> currently the frame-name is calculated based on the labels of the value-field. this will have to change somehow. i think Prometheus has the same problem

@gabor good call, i'll update this

> for instant-queries, we post-process the results a lot in the loki-datasource-in-javascript to form a table-result, this code will need to get updated (maybe we can re-use some code from some other grafana-javascript)

looking at the table results, it looks like it's a long time series. maybe the frontend can switch to calling the prepare time series wide -> long transformation?

> in general, this change is backward-incompatible.i mean, this is observable by the end-user if they use any transforms. so we have to make a decision whether we do it or not. if we do it, grafana9 would be the best time to do it, as that's a major version, where breaking-changes are the most accepted. but maybe it's too little time for this change? don't know.

posted this above, but we could make this change backwards compatible by setting a value only on new queries that indicate that they should return wide frames. existing and edited queries wouldn't have this property, so they would return multi/many. thoughts?

gabor commented:
@toddtreece 

> > in general, this change is backward-incompatible.i mean, this is observable by the end-user if they use any transforms. so we have to make a decision whether we do it or not. if we do it, grafana9 would be the best time to do it, as that's a major version, where breaking-changes are the most accepted. but maybe it's too little time for this change? don't know.
> 
> posted this above, but we could make this change backwards compatible by setting a value only on new queries that indicate that they should return wide frames. existing and edited queries wouldn't have this property, so they would return multi/many. thoughts?

i agree that it could work, but i'd personally prefer not to do it for 2 reasons:
1. this would cause that we need to have two implementations (wide and not-wide) for a long time in parallel, because we cannot remove the not-wide-code.
2. this can get fragile in a dashboard. imagine:
    - you create a dashboard with a query (while in old-mode), save it, all runs.
    - we release the wide-thing, dashboard still runs old-queries, all runs.
    - now you edit the dashboard, add one more query, and suddenly the result-data has a different format. i feel like that would be very surprising to the user
    - also, when editing an existing query, we have to decide if we "upgrade" in that case or no

toddtreece commented:
> this would cause that we need to have two implementations (wide and not-wide) for a long time in parallel, because we cannot remove the not-wide-code.

unfortunately i think this is the case for all changes like this at the moment. i'll bring it up in the data plane channel

toddtreece committed:
Merge branch 'main' of github.com:grafana/grafana into toddtreece/prom-converter-wide-frame

toddtreece committed:
update tests

toddtreece committed:
conditionally enable wide response

toddtreece committed:
add prometheusWideSeries feature flag

toddtreece committed:
fix lint err

toddtreece committed:
use feature flag in frame/field naming

toddtreece commented:
@gabor @ryantxu i added a feature flag for prometheus, and disabled wide frames for now for loki. i think this PR should be ready for 👀 if anyone has time

toddtreece committed:
Merge branch 'main' of github.com:grafana/grafana into toddtreece/prom-converter-wide-frame

gabor commented:
@toddtreece , from a quick check, code looks ok:

1. the loki-related part is OK
2. the wide-frame-generation-code looks OK too (though somebody better at `jsoniter` should probably review that part)

toddtreece committed:
Merge branch 'main' of github.com:grafana/grafana into toddtreece/prom-converter-wide-frame

ryantxu committed:
Merge remote-tracking branch 'origin/main' into toddtreece/prom-converter-wide-frame

ryantxu commented on pkg/tsdb/prometheus/testdata/range_simple.result.json:
```diff
@@ -25,7 +25,6 @@
         },
         "values": [
           [1641889530.123, "54"],
-          [1641889531.123, "65"],
```
where did it go?

toddtreece replied:
@ryantxu i think this was removed to simulate a gap in one of the series

ryantxu reviewed (approved):
Looks good!   lets merge and get the feature toggle set in ops!

ryantxu committed:
Merge remote-tracking branch 'origin/main' into toddtreece/prom-converter-wide-frame
